### PR TITLE
Decrease preallocated partition container sizes

### PR DESCRIPTION
**Pull request by mmedenjak**
https://github.com/hazelcast/hazelcast-enterprise/pull/3536

Code to test effects - adds a single entry in each of 20k partitions, clears the map and then we take a heap dump to see the size:
```java
public static void main(String[] args) throws InterruptedException {
        int partitions = 20_001;
        Config c = new Config()
                .setLicenseKey(licenseKey)
                .setProperty(ClusterProperty.PARTITION_COUNT.getName(), String.valueOf(partitions));

        HazelcastInstance i = Hazelcast.newHazelcastInstance(c);
        IMap<Object, Object> map = i.getMap("map");
        IMap<Object, Object> map2 = i.getMap("map1");

        int parallelism = 4;
        Thread[] threads = new Thread[parallelism];

        for (int j = 0; j < parallelism; j++) {
            int finalJ = j;
            Thread t = new Thread(() -> {
                for (int k = finalJ; k < partitions; k += parallelism) {
                    String key = generateKeyForPartition(i, finalJ);
                    map.put(key, "");
                    map2.put(key, "");
                }
            });
            t.start();
            threads[j] = t;
        }
        for (Thread thread : threads) {
            thread.join();
        }

        System.out.println("DONE " + map.size() + " " + map2.size());
        map.clear();
        map2.clear();
        System.out.println("DONE 2 " + map.size() + " " + map2.size());

    }
```

Heap dump after code runs, with this PR: 312M
Heap dump after code runs, on master: 616M

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java [hunk 1]
```diff
@@ -103,7 +104,7 @@
     private static final Object NULL_OBJECT = new Object();
 
     private final ConcurrentMap<String, QueueContainer> containerMap = new ConcurrentHashMap<>();
-    private final ConcurrentMap<String, LocalQueueStatsImpl> statsMap = new ConcurrentHashMap<>(1000);
+    private final ConcurrentMap<String, LocalQueueStatsImpl> statsMap = MapUtil.createConcurrentHashMap(10);
```
Comment (blazember) [inline hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: Can't we initialize this with the number of DSs defined in the static config? Then we can grow on-demand if new ones created dynamically.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: yes, this was also an idea, have to see how (un)intrusive it is to get the config in the constructor and iterate through it.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: wildcard configs are a gotcha but I think it's fine to work with best-effort approach here.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/20/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/20/Hazelcast-pr-builder-20.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[44,24] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[49,8] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[67,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[69,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[71,70] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[73,23] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,30] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,45] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[98,44] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java:[51,60] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[44,24] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[49,8] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[67,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[69,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[71,70] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[73,23] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,30] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,45] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[98,44] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java:[51,60] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java:[52,59] error: cannot find symbol
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/22/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/22/Hazelcast-pr-builder-22.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 3 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapPutAllMessageTask.java:50: Line is longer than 130 characters (found 153). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/LocalMapStatsProvider.java:48:8: Unused import - java.util.concurrent.ConcurrentHashMap. [UnusedImports]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java:42:8: Unused import - java.util.concurrent.ConcurrentHashMap. [UnusedImports]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/976/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/976/Hazelcast-pr-builder-976.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientReliableTopicOnClusterRestartTest.shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant:185->lambda$shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant$5:187
[INFO] 
[ERROR] Tests run: 27745, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: Addressed this comment, also on other places. The approach is pretty simple, I just take the number of configs, which means wildcard and default also count as one. PTAL

**Comment by mmedenjak** [comment]
Merging with unrelated test failure, thank you for the reviews, everyone!

**Pull request merged**